### PR TITLE
Added UsesJavaVersion to filter applicability by java version.

### LINKED ISSUE #983: Add UsesJavaVersion visitor to filter applicability by target version of java

### PR DESCRIPTION
**Pull request by traceyyoshima**
Added UsesJavaVersion visitor.
Added MajorVersion enum to JavaVersion.
fixes #983

### CODE REVIEW

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java [hunk 1]
```diff
@@ -43,4 +34,57 @@
...
     String sourceCompatibility;
     String targetCompatibility;
+
+    public static MajorVersion getMajorVersion(String version) {
+        return MajorVersion.get(version);
+    }
+
+    public enum MajorVersion {
+        INVALID("invalid"),
+        JAVA_6("6"),
```
Comment (jkschneider) [inline rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:44]: Why this enum?

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java [hunk 2]
```diff
@@ -43,4 +34,57 @@
...
+        JAVA_10("10"),
+        JAVA_11("11"),
+        JAVA_12("12"),
+        JAVA_13("13"),
+        JAVA_14("14"),
+        JAVA_15("15"),
+        JAVA_16("16"),
+        JAVA_17("17");
+
+        private static final Map<String, MajorVersion> lookup = new HashMap<>();
```
Comment (tkvangorder) [inline rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:57]: I am not sure you even need a map here, we are talking about iterating over 11 values.

I would just iterate over those values: 

```
public static MajorVersion get(String version) {
    for (MajorVersion v : MajorVersion.values()) {
        if (v.value.equals(version)) {
            return v;
        }
    }
    return INVALID;
}
```

Comment (tkvangorder) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:57]: If you wanted to, you could order the enums from most popular to least, but I just don't feel this micro-optimization buys you much.

**Review by tkvangorder — CHANGES_REQUESTED** [review]
Just a small change in your enum, otherwise, looks fine to me.

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java (hunk 1 above)
Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:44]: I waited for feedback on the enum before adding another constructor to UsesJavaVersion.

We can have a method like `String getMajorVersion(String)` that normalizes the input, but I think an enum is clearer for recipe authors: `new UsesJavaVersion(JavaVersion.MajorVersion.Java_8)`.

Do you have concerns about an enum?

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java (hunk 2 above)
Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:57]: Fair point, thanks! I added a response to Jon's question about the enum, and will add that based on his response.

Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java]: Add the changes

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java (hunk 1 above)
Comment (jkschneider) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:44]: Why not just an integer? `8`, `11`? Just feels like bloat with little benefit.

Comment (tkvangorder) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java:44]: hah, now looking at the enum in the context of this discussion, An integer makes quite a lot of sense.

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java (hunk 2 above)
Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java]: Update removed the enum per Jon's comment

File: rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java (hunk 1 above)
Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/marker/JavaVersion.java]: Updated to use an int

**Review by tkvangorder — APPROVED** [review]
LGTM

**Pull request merged**